packages/client/lib/client/authx/token-manager.spec.ts

```diff
@@ -0,0 +1,614 @@
+
+          assert.equal(manager.getCurrentToken(), null, 'Should not have token yet');
+          // Wait for the first token request to complete ( it should be immediate, and we should wait only for the network delay)
+          await delay(networkDelay)
+
+          assert.equal(listener.receivedTokens.length, 1, 'Should receive initial token');
+          assert.equal(listener.receivedTokens[0].value, 'token1', 'Should have correct token value');
+          assert.equal(listener.receivedTokens[0].expiresAtMs, 100, 'Should have correct expiration time');
+          assert.equal(listener.receivedTokens[0].receivedAtMs, 0, 'Should have correct receivedAt time');
+          assert.equal(listener.errors.length, 0, 'Should not have any errors: ' + listener.errors);
+          assert.equal(manager.getCurrentToken().value, 'token1', 'Should have current token');
+
+          clock.advance(80);
+          await delay(80);
+
+          assert.equal(listener.receivedTokens.length, 1, 'Should not receive new token yet');
+          assert.equal(listener.errors.length, 0, 'Should not have any errors');
+
+          clock.advance(networkDelay);
+          await delay(networkDelay);
+
+          assert.equal(listener.receivedTokens.length, 2, 'Should receive second token');
+          assert.equal(listener.receivedTokens[1].value, 'token2', 'Should have correct token value');
+          assert.equal(listener.receivedTokens[1].expiresAtMs, 200, 'Should have correct expiration time');
+          assert.equal(listener.receivedTokens[1].receivedAtMs, 100, 'Should have correct receivedAt time');
+          assert.equal(listener.errors.length, 0, 'Should not have any errors');
+          assert.equal(manager.getCurrentToken().value, 'token2', 'Should have current token');
+
+          clock.advance(80);
+          await delay(80);
+
+          assert.equal(listener.receivedTokens.length, 2, 'Should not receive new token yet');
+          assert.equal(listener.errors.length, 0, 'Should not have any errors');
+
+          clock.advance(networkDelay);
+          await delay(networkDelay);
+
+          assert.equal(listener.receivedTokens.length, 3, 'Should receive third token');
+          assert.equal(listener.receivedTokens[2].value, 'token3', 'Should have correct token value');
+          assert.equal(listener.receivedTokens[2].expiresAtMs, 300, 'Should have correct expiration time');
+          assert.equal(listener.receivedTokens[2].receivedAtMs, 200, 'Should have correct receivedAt time');
+          assert.equal(listener.errors.length, 0, 'Should not have any errors');
+          assert.equal(manager.getCurrentToken().value, 'token3', 'Should have current token');
+
+          disposable.dispose();
+        });
+      });
+    });
+  });
+
+  describe('TokenManager error handling', () => {
+
+    describe('error scenarios', () => {
+      it('should not recover if retries are not enabled', async () => {
+        const clock = new FakeClock(0);
```

Comment:
node allows [synthetic clocks](https://nodejs.org/docs/latest-v18.x/api/test.html#timers), so it may be preferred to use them over a custom faked clock?

Reply:
Thanks @dhensby! You're the second node developer to suggest using node's synthetic timers. While I have a slight preference for avoiding global method replacements like setTimeout/Date.now() , you're right - it would be better to align with common practices. I'll switch to using Node's synthetic timers instead of my custom Clock approach.